polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -789,362 +665,335 @@ impl FragmentTree {
+	}
+
+	// Return the earliest relay parent a new candidate can have in order to be added to the chain.
+	// This is the relay parent of the last candidate in the chain.
+	// The value returned may not be valid if we want to add a candidate pending availability, which
+	// may have a relay parent which is out of scope. Special handling is needed in that case.
+	// `None` is returned if the candidate's relay parent info cannot be found.
+	fn earliest_relay_parent(&self) -> Option<RelayChainBlockInfo> {
+		if let Some(last_candidate) = self.chain.last() {
+			self.scope.ancestor(&last_candidate.relay_parent()).or_else(|| {
+				// if the relay-parent is out of scope _and_ it is in the chain,
+				// it must be a candidate pending availability.
+				self.scope
+					.get_pending_availability(&last_candidate.candidate_hash)
+					.map(|c| c.relay_parent.clone())
+			})
+		} else {
+			Some(self.scope.earliest_relay_parent())
+		}
 	}
 
-	// Orders the ancestors into a viable path from root to the last one.
-	// Returns a pointer to the last node in the path.
-	// We assume that the ancestors form a chain (that the
-	// av-cores do not back parachain forks), None is returned otherwise.
-	// If we cannot use all ancestors, stop at the first found hole in the chain. This usually
-	// translates to a timed out candidate.
-	fn find_ancestor_path(&self, mut ancestors: Ancestors) -> Option<NodePointer> {
-		// The number of elements in the path we've processed so far.
-		let mut depth = 0;
-		let mut last_node = NodePointer::Root;
-		let mut next_node: Option<NodePointer> = Some(NodePointer::Root);
-
-		while let Some(node) = next_node {
-			if depth > self.scope.max_depth {
-				return None;
+	// Checks if this candidate could be added in the future to this chain.
+	// This assumes that the chain does not already contain this candidate. It may or may not be
+	// present in the `CandidateStorage`.
+	// Even if the candidate is a potential candidate, this function will indicate that it can be
+	// kept only if there's enough room for it.
+	pub(crate) fn can_add_candidate_as_potential(
+		&self,
+		storage: &CandidateStorage,
+		candidate_hash: &CandidateHash,
+		relay_parent: &Hash,
+		parent_head_hash: Hash,
+		output_head_hash: Option<Hash>,
+	) -> PotentialAddition {
+		// If we've got enough candidates for the configured depth, no point in adding more.
+		if self.chain.len() > self.scope.max_depth {
+			return PotentialAddition::None
+		}
+
+		if !self.check_potential(relay_parent, parent_head_hash, output_head_hash) {
+			return PotentialAddition::None
+		}
+
+		let present_in_storage = storage.contains(candidate_hash);
+
+		let unconnected = self
+			.find_unconnected_potential_candidates(
+				storage,
+				present_in_storage.then_some(candidate_hash),
+			)
+			.len();
+
+		if (self.chain.len() + unconnected) < self.scope.max_depth {
+			PotentialAddition::Anyhow
+		} else if (self.chain.len() + unconnected) == self.scope.max_depth {
```

Comment:
Why if we are already at max_depth we can add one more ?

Reply:
that's a good question.

Even prior to this PR, we were accepting max_depth + 1 unincluded candidates at all times.
Not sure why it was designed this way. I assume it's because a max_candidate_depth of 0 is equal to synchronous backing and would still allow for one candidate to be backed (therefore the +1 is needed)

Reply:
I see that makes some sense indeed.

Reply:
The question is that now we have async backing enabled do we still want to keep this ?

Reply:
> The question is that now we have async backing enabled do we still want to keep this ?

I don't have a good reason for chaning this. Moreover, we seem to have discovered that we need a max_depth + 1 unincluded segment size also on the collator. So maybe it's here for a reason.
Nevertheless, I'd leave it as is unless we find a good reason to change it